docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+rejected.
+
+Bluebird also has a `Promise.race()` function which does the same thing but
+doesn't wait for all of them to finish. That is what we want. An example how
+this might look:
+
+```javascript
+function showProgress(otherPromise) {
+  var progress = new ProgressbarDialog().show('Uploading…');
+  return Promise.race([otherPromise, promise.waitForUser()])
+    .finally(function() {
+      progress.hide();
+    });
+}
+```
+
+Here is some example HTML for the Progress Dialog:
+
+```html
+<style type="text/css">
+  #progress-dialog {
+    width:      200px;
+    margin:     auto;
+    border:     thin solid black;
+    padding:    10px;
+    background: lightgreen;
+  }
+  #progress-dialog .progress-bar {
+    border:  1px solid black;
+    margin:  10px auto;
+    padding: 0;
+    height:  20px;
+  }
+  #progress-dialog .progress-bar>div {
+    background-color: blue;
+    margin:           0;
+    padding:          0;
+    border:           none;
+    height:           20px;
+  }
+</style>
+<div id="progress-dialog">
+  <div class="message"></div>
+  <div class="progress-bar"><div></div></div>
+  <div>
+    <button class="cancel">Cancel</button>
+  </div>
+</div>
+```
+
+The JavaScript is the same as the `PromptDialog` only we will add a
+`setProgress()` method:
+
+```javascript
+function ProgressDialog() {
+  Dialog.call(this);
+  this.el           = document.getElementById('progress-dialog');
+  this.messageEl    = this.el.querySelector('.message');
+  this.progressBar  = this.el.querySelector('.progress-bar>div');
+  this.cancelButton = this.el.querySelector('button.cancel');
+  this.attachDomEvents();
+}
+ProgressDialog.prototype = Object.create(Dialog.prototype);
+ProgressDialog.prototype.attachDomEvents = function() {
+  var _this = this;
+  this.cancelButton.addEventListener('click', function() {
+    _this.cancel();
+  });
+  
+};
+ProgressDialog.prototype.show = function(message) {
+  this.messageEl.innerHTML = '' + message;
+  this.el.className = '';
+  return this;
+};
+ProgressDialog.prototype.hide = function() {
+  this.el.className = 'hidden';
+  return this;
+};
+ProgressDialog.prototype.setProgress = function(percent) {
+  this.progressBar.style.width = percent + '%';
+};
+```
+
+A common misconception is that a promises a form of callback management. This
+is not the case and is why the idea of having a progress callback is not part of
+the Promise spec. However, much like the Promise library passes in a `resolve`
+and `reject` callback when you create a new promise (`new Promise(…)`) we can do
+the same patter for a progress callback.
+
+Now to the fun part. For this tutorial we will *fake* a lengthy file upload by
+using `setTimeout`. The intent is to provide a promise and to allow a progress
+to be periodically ticked away. We will expect a function to be passed which
+is called whenever the progress needs updating. And it returns a promise.
+
+```javascript
+function delayedPromise(progressCallback) {
```

Comment:
It's important to illustrate that we're returning a promise _and_ taking a callback here - it's also important to talk about what happens if the callback throws.